nntrainer/compiler/slice_realizer.cpp

```diff
@@ -53,39 +56,38 @@ SliceRealizer::realize(const GraphRepresentation &reference) {
   auto cur_start_layers = start_layers;
   auto cur_end_layers = end_layers;
 
-  if (start_layers.empty()) {
-    for (auto &node : reference) {
-      if (node->getNumInputConnections() == 0) {
-        cur_start_layers.push_back(node->getName());
+  /** setup children before filling in the end layers */
+  std::for_each(reference.begin(), reference.end(),
+                [&mp](std::shared_ptr<LayerNode> node) {
+                  auto node_name = node->getName();
+                  for (auto &parent : node->getInputLayers()) {
+                    mp.at(parent).children.push_back(node_name);
+                  };
+                });
+
+  if (cur_start_layers.empty()) {
+    for (auto &node : mp) {
+      if (node.second.node->getNumInputConnections() == 0) {
+        cur_start_layers.push_back(node.second.node->getName());
       }
     }
   }
 
-  if (end_layers.empty()) {
+  if (cur_end_layers.empty()) {
     for (auto &node : mp) {
       if (node.second.children.size() == 0) {
         cur_end_layers.insert(node.first);
       }
     }
   }
 
-  std::for_each(reference.begin(), reference.end(),
-                [&mp](std::shared_ptr<LayerNode> node) {
-                  auto node_name = node->getName();
-                  for (auto &parent : node->getInputLayers()) {
-                    mp.at(parent).children.push_back(node_name);
-                  };
-                });
-
-  GraphRepresentation processed;
+  if (cur_start_layers.empty()) {
+    throw std::runtime_error("No start layer is found, graph has a loop.");
```

Comment:
nitpicking: even if start/end layers are not empty, it is possible that there is a loop, wonder if this is considered?

Reply:
No, that isn't checked. That check should be done outside the slice realizer.